state.go

```diff
@@ -0,0 +1,308 @@
+		BaseState{
+			id: MenderStateUpdateFetch,
+		},
+		update,
+	}
+}
+
+func (u *UpdateFetchState) Handle(c Controler) (State, bool) {
+	log.Debugf("handle update fetch state")
+	in, size, err := c.FetchUpdate(u.update.Image.URI)
+	if err != nil {
+		log.Errorf("update fetch failed: %s", err)
+		return NewErrorState(NewTransientError(err)), false
+	}
+
+	return NewUpdateInstallState(in, size), false
+}
+
+type UpdateInstallState struct {
+	BaseState
+	// reader for obtaining image data
+	imagein io.ReadCloser
+	// expected image size
+	size int64
+}
+
+func NewUpdateInstallState(in io.ReadCloser, size int64) State {
+	return &UpdateInstallState{
+		BaseState{
+			id: MenderStateUpdateInstall,
+		},
+		in,
+		size,
+	}
+}
+
+func (u *UpdateInstallState) Handle(c Controler) (State, bool) {
+	log.Debugf("handle update install state")
+	if err := c.InstallUpdate(u.imagein, u.size); err != nil {
+		log.Errorf("update install failed: %s", err)
+		return NewErrorState(NewTransientError(err)), false
+	}
+
+	if err := c.EnableUpdatedPartition(); err != nil {
+		log.Errorf("enabling updated partition failed: %s", err)
+		return NewErrorState(NewTransientError(err)), false
+	}
+
+	return rebootState, false
+}
+
+type UpdateCheckWaitState struct {
+	BaseState
+	cancel chan (bool)
+}
+
+func NewUpdateCheckWaitState() State {
+	return &UpdateCheckWaitState{
+		BaseState{
+			id: MenderStateUpdateCheckWait,
+		},
+		make(chan bool),
+	}
+}
+
+func (u *UpdateCheckWaitState) Handle(c Controler) (State, bool) {
+	log.Debugf("handle update check wait state")
+
+	intvl := c.GetUpdatePollInterval()
+	log.Debugf("wait %v before next poll", intvl)
+	ticker := time.NewTicker(intvl)
+
+	select {
+	case <-ticker.C:
+		log.Debugf("poll interval complete")
+		return updateCheckState, false
+	case <-u.cancel:
+		log.Infof("update wait canceled")
+	}
+
+	return u, true
+}
+
+// Cancel wait state
+func (u *UpdateCheckWaitState) Cancel() bool {
+	u.cancel <- true
+	return true
+}
+
+type ErrorState struct {
+	BaseState
+	cause menderError
+}
+
+func NewErrorState(err menderError) State {
+	if err == nil {
+		err = NewFatalError(errors.New("general error"))
+	}
+
+	return &ErrorState{
+		BaseState{
+			id: MenderStateError,
+		},
+		err,
+	}
+}
+
+func (e *ErrorState) Handle(c Controler) (State, bool) {
+	log.Infof("handling error state, current error: %v", e.cause.Error())
+	// decide if error is transient, exit for now
+	if e.cause.IsFatal() {
+		return doneState, false
+	}
+	return initState, false
+}
+
+func (e *ErrorState) IsFatal() bool {
+	return e.cause.IsFatal()
+}
+
+type RebootState struct {
+	BaseState
+}
+
+func (e *RebootState) Handle(c Controler) (State, bool) {
+	log.Debugf("handle reboot state")
+	if err := c.Reboot(); err != nil {
+		return NewErrorState(NewFatalError(err)), false
+	}
+	return doneState, false
+}
+
+type FinalState struct {
+	BaseState
+}
+
+func (f *FinalState) Handle(c Controler) (State, bool) {
+	panic("reached final state")
```

Comment:
That's the point. This is the final state, you should not execute as we're technically done. See what daemon code with it: https://github.com/mendersoftware/mender/pull/68/files/2139bf44a7ba4e67dd724ab7d5bd94e2803ef21c#diff-e7bcc6178f83a5063b3c27cc5a860780R55